### PR TITLE
Student card

### PR DESCRIPTION
**Pull request by dziugas306**

### CODE REVIEW

File: src/app/admin-page/admin-page.component.html [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <div class="spaced-sides">
   <div class="row">
     <div class="col-4">
-      <app-student-list></app-student-list>
+      <app-student-list [isInAdminOrManger] = "true"></app-student-list>
```
Comment (GVabal) [inline src/app/admin-page/admin-page.component.html]: typo in Manager

**Review by drachno — APPROVED** [review]

File: src/app/manager-page/manager-page.component.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<div class="spaced-sides">
+    <div class="row">
+        <div class="col-4">
+            <app-student-list [isInAdminOrManger] = "false"></app-student-list>
```
Comment (pauliusnet) [inline src/app/manager-page/manager-page.component.html]:
```suggestion
            <app-student-list [isInAdminOrManager] = "false"></app-student-list>
```

File: src/app/manager-page/student-card/student-card.component.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<div *ngIf="this.student" class="mt-5 shadow bg-white rounded">
+  <div class="d-flex justify-content-between">
+    <img class="m-4"src="{{student.pictureUrl}}" style="border-radius:50%;" width="300" height="300">
+    <div class="">
```
Comment (pauliusnet) [inline src/app/manager-page/student-card/student-card.component.html]:
```suggestion
    <div>
```

File: src/app/manager-page/student-card/student-card.component.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<div *ngIf="this.student" class="mt-5 shadow bg-white rounded">
+  <div class="d-flex justify-content-between">
+    <img class="m-4"src="{{student.pictureUrl}}" style="border-radius:50%;" width="300" height="300">
```
Comment (pauliusnet) [inline src/app/manager-page/student-card/student-card.component.html]: maybe move this styling to ```student-card.component.scss```?
```suggestion
    <img class="m-4"src="{{student.pictureUrl}}">
```

File: src/app/manager-page/student-card/student-card.component.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+@Component({
+  selector: 'app-student-card',
+  templateUrl: './student-card.component.html',
+  styleUrls: ['./student-card.component.scss']
+})
+export class StudentCardComponent implements OnInit {
+
+  studentId$!: Observable<number>;
+  student!: Student;
```
Comment (pauliusnet) [inline src/app/manager-page/student-card/student-card.component.ts]:
```suggestion
  student$!: Observable<Student>;
```

File: src/app/manager-page/student-card/student-card.component.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  constructor(private store: Store) { }
+
+  ngOnInit(): void {
+    this.studentId$ = this.store.select(getSelectedStudentId);
+    this.studentId$.pipe(
+      switchMap(studentId => (this.store.select(getStudentById(studentId)) as Observable<Student>))
+    ).subscribe((student) => {
+      if (student) {
+        this.student = student;
+    }});
```
Comment (pauliusnet) [inline src/app/manager-page/student-card/student-card.component.ts]:
```suggestion
    this.student$ = this.store.pipe(
      select(getSelectedStudentId),
      switchMap(studentId => (this.store.select(getStudentById(studentId)) as Observable<Student>))
```

I think you can you use observable straight away :)

File: src/app/manager-page/student-card/student-card.component.html [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<div *ngIf="this.student" class="mt-5 shadow bg-white rounded">
```
Comment (pauliusnet) [inline src/app/manager-page/student-card/student-card.component.html]:
```suggestion
<div *ngIf="this.student$ | async as student" class="mt-5 shadow bg-white rounded">
```

File: src/app/manager-page/student-card/student-card.component.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  average(list: number[]): number{
+    let sum = 0;
+    for (const item of list ){
+      sum += item;
+    }
+    const avg = sum / list.length;
+    return Number(avg.toFixed(2));
+  }
+}
```
Comment (pauliusnet) [inline src/app/manager-page/student-card/student-card.component.ts]: what about moving this logic to selector and writing a new selector, that would return averageReviewValues :)

File: src/app/service/review.service.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ReviewService {
+
+  constructor(private http: HttpClient) { }
+
+  getReviewsByStudentId(id: number): Observable<Review[]>{
+    return this.http.get<Review[]>(apiUrl + '/' + id + '/reviews');
```
Comment (pauliusnet) [inline src/app/service/review.service.ts]:
```suggestion
    return this.http.get<Review[]>(apiUrl + `/${id}/reviews`);
```

File: src/app/shared-components/radar-chart/radar-chart.component.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      }
+  },
+  legend: {
+    display: false
+  }
+  };
+
+  ngOnInit(): void {
+  }
+}
```
Comment (pauliusnet) [inline src/app/shared-components/radar-chart/radar-chart.component.ts]:
```suggestion
}
```
you can remove this ant the ```implements OnInit```, if it is not used :)

File: src/app/shared-components/student-list/student-list.component.ts [hunk 11]
```diff
@@ -18,6 +18,7 @@ export class StudentListComponent implements OnInit {
   isStudentsLoaded$: Observable<boolean> | undefined ;
   hasLoadFailed$: Observable<boolean> | undefined ;
   error$: Observable<CustomError | null> | undefined;
+  @Input() isInAdminOrManger: boolean | undefined;
```
Comment (pauliusnet) [inline src/app/shared-components/student-list/student-list.component.ts]:
```suggestion
  @Input() isManagerView: boolean | undefined;
```
Naming is a bit unclear, so if it's true, does it make admin or manager? Try renaming to this :)

File: src/app/store/review/review.selectors.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+import { mapToReviewData } from 'src/app/shared/review';
+import { reviewsAdapter, reviewsFeatureKey, ReviewsState } from './review.reducer';
+
+
+export const getReviewsFeatureState =  createFeatureSelector<ReviewsState>(reviewsFeatureKey);
+export const getIsReviewsLoading = createSelector(getReviewsFeatureState, reviewsState => reviewsState.loading);
+export const getIsReviewsLoaded = createSelector(getReviewsFeatureState, reviewsState => reviewsState.loaded);
+export const getReviewsError = createSelector(getReviewsFeatureState, reviewsState => reviewsState.error);
+export const {selectAll: selectReviews} = reviewsAdapter.getSelectors(getReviewsFeatureState);
+export const getReviewData =  createSelector(selectReviews, (reviews) =>  mapToReviewData(reviews));
```
Comment (pauliusnet) [inline src/app/store/review/review.selectors.ts]:
```suggestion
export const getReviewData =  createSelector(selectReviews, (reviews) =>  mapToReviewData(reviews));

export const getReviewDataAverages = createSelector(getReviewData, (data) => ([average(data.overallGrade), average(data.abilityToLearnGrade), average(data.motivationGrade),
         average(data.extraMileGrade), average(data.communicationGrade)]);
         
 function average(list: number[]): number{
    let sum = 0;
    for (const item of list ){
      sum += item;
    }
    const avg = sum / list.length;
    return Number(avg.toFixed(2));
  }
```

File: src/app/store/students/students.reducer.ts [hunk 13]
```diff
@@ -19,6 +19,7 @@ export interface StudentsState extends EntityState<Student> {
   hasStudentDeleteFailed: boolean;
   error: CustomError | null;
   studentEditId: number;
+  selectedStudentId: number;
```
Comment (pauliusnet) [inline src/app/store/students/students.reducer.ts]:
```suggestion
  selectedStudentId?: number;
```

File: src/app/store/students/students.reducer.ts [hunk 14]
```diff
@@ -34,6 +35,7 @@ export const initialState: StudentsState = studentsAdapter.getInitialState({
   hasStudentLoadFailed: false,
   editOrCreateForm: false,
   hasStudentDeleteFailed: false,
+  selectedStudentId: -1,
```
Comment (pauliusnet) [inline src/app/store/students/students.reducer.ts]:
```suggestion
```

File: src/app/store/students/students.reducer.ts [hunk 15]
```diff
@@ -127,11 +129,16 @@ export const studentsReducer = createReducer(
...
+    let selectedStudentId = -1;
+    if (action.students[0]){
+      selectedStudentId = action.students[0].id;
+    }
     return studentsAdapter.addMany(action.students, {
       ...state,
       loading: false,
       loaded: true,
-      error: null
+      error: null,
```
Comment (pauliusnet) [inline src/app/store/students/students.reducer.ts:138]:
```suggestion
```

**Review by pauliusnet — CHANGES_REQUESTED** [review]
Some changes needed to proceed but all in all - very good work!! 💯

**Review by GVabal — APPROVED** [review]

**Pull request merged**